### PR TITLE
Stop creating unused variables

### PR DESCRIPTION
**Pull request by 0dminnimda**

### CODE REVIEW

**Comment by 0dminnimda** [comment]
Is it reasonable first try to merge it to cpython?

**Comment by pablogsal** [comment]
> Is it reasonable to first try to merge it to cpython?

Generally we do it the other way: we merge stuff here and I sync them from time to time in batches.

**Comment by 0dminnimda** [comment]
Ok, I see.
Also can you ask for the launch of the workflow?

**Comment by 0dminnimda** [comment]
Probably a good time to say, that two tests are failing on the `main`. The same tests are failing here. Although I think fixing those tests is out of the scope of this pr, unfortunately

**Comment by MatthieuDartiailh** [comment]
Getting #66 in first would make sense to get CIs to test on more versions.

**Comment by 0dminnimda** [comment]
Does #66 fixes broken tests?

**Comment by 0dminnimda** [comment]
> Generally we do it the other way: we merge stuff here and I sync them from time to time in batches

Well, then what's about distributing credit/blame for the changes? Will those changes be cherry-picked into cpython?

**Comment by pablogsal** [comment]
> > Generally we do it the other way: we merge stuff here and I sync them from time to time in batches
> 
> Well, then what's about distributing credit/blame for the changes? Will those changes be cherry-picked into cpython?

No, they will be directly included and blame/credit is lost.

**Comment by 0dminnimda** [comment]
> No, they will be directly included in one batch and blame/credit is lost.

Aww, isn't this a reason to make a direct pr to cpython?

Well, not like it's *that* important, but it would be nice if the credit/blame was preserved!

**Comment by MatthieuDartiailh** [comment]
@0dminnimda Can you rebase on main so that we get an extended test run ?

**Comment by 0dminnimda** [comment]
Oh yeah, workflow activation, what a lovely github feature ♥

**Comment by gvanrossum** [comment]
Bye, I am leaving this repo now.

**Comment by 0dminnimda** [comment]
oh .. bye

<br>
<br>
<br>

well, we need to get back to the work

3.10 didn't pass, it seems like versoin of python that tox runs is somehow different from the other ones
Because if I try replicate this I see a string without `'`

I don't quite understand why it happens

**Comment by MatthieuDartiailh** [comment]
#71 will fix the failure on 3.10 (3.10.5 fixed an issue for which we had a workaround)

**Comment by 0dminnimda** [comment]
Closing and opening to triger the ci

**Comment by 0dminnimda** [comment]
oh yeah, forgot to merge

**Comment by 0dminnimda** [comment]
Is there any additional comments or this can be merged?

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [inline src/pegen/python_generator.py:201]: I tend to find that kind of aliasing to be detrimental to the readability personally (in particular for such a simple alias). But I will let @pablogsal comment on this.

File: src/pegen/grammar_parser.py
Comment (MatthieuDartiailh) [inline src/pegen/grammar_parser.py:240]: This looks wrong to me. Updating cut has side effects (l 268) and they are not taken into account.

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
I found one concerning issue regarding cut (and I find even more concerning that no test found the issue). I also left a couple of comment/question.

File: src/pegen/grammar_parser.py
Comment (0dminnimda) [reply src/pegen/grammar_parser.py:240]: oh yeah, I didn't handle cut properly

File: src/pegen/python_generator.py
Comment (0dminnimda) [reply src/pegen/python_generator.py:352]: Well, it raises another issue, what if the LOCATIONS formatting does not use keyword args as are now set bu the default. In this case this will break

Comment (0dminnimda) [reply src/pegen/python_generator.py:352]: I'm gonna try to make formating apply before the usednamesvisitor

**Comment by 0dminnimda** [comment]
Maybe @pablogsal have other comments?

File: src/pegen/python_generator.py
Comment (0dminnimda) [reply src/pegen/python_generator.py]: Also, @MatthieuDartiailh do you have any comments regarding new implementation? 
If no, you can resolve this convo.

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [inline src/pegen/python_generator.py:399]: No need for the 2 ifs

```suggestion
        locations = False
        unreachable = False
        used = None
        if action:
            if "LOCATIONS" in action:
                locations = True
                action = action.replace("LOCATIONS", self.location_formatting)
            if "UNREACHABLE" in action:
                unreachable = True
                action = action.replace("UNREACHABLE", self.unreachable_formatting)

            used = self.usednamesvisitor.visit(ast.parse(action))
            if has_cut:
                used.add("cut")
```

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [inline src/pegen/python_generator.py:333]: Could you just point me to a concrete example of this issue ?

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
I left a minor suggestions for improvement and a question otherwise LGTM.

File: src/pegen/python_generator.py
Comment (0dminnimda) [reply src/pegen/python_generator.py:399]: oh, yeah! right 😁

Comment (0dminnimda) [reply src/pegen/python_generator.py:399]: Well, actually .. I think the original version is more readable. Not like the second if will lead to critical performance hit.
Here we have two if's for different logic, easy to follow and change

Comment (0dminnimda) [reply src/pegen/python_generator.py:399]: Idk, I'll left it for the final review by the pablogsal

File: src/pegen/python_generator.py
Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: Woulld love to!
You cannot find it in the metaparser, but in the python parser.

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: For example:
```python
@memoize
    def file(self) -> Optional[ast . Module]:
        # file: statements? $
        mark = self._mark()
        if (
            (a := self.statements(),)
            and
            (self.expect('ENDMARKER'))
        ):
            return ast . Module ( body = a or [] , type_ignores = [] );
        self._reset(mark)
        return None;
```

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: As it seems all of the rules like `X?` will result in this

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: And `[X]` will also result in this

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: It seems logical, optional rule cannot be incorrect as if there's nothing found it's still fine!

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:399]: IMO having a single if with a comment before the inner ifs and before the systematic part would be just as readable without duplicating the if test.

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:333]: I need to look at how things work currently and why this is now needed.

Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:333]: So I checked and in the case of the file rule the parentheses are already there for the case of statements so I believe we do not need this change or am I missing something ?

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: parentheses are there because the name is used so
`self.print(f"({name} := {call})")` is responsible
But if we don't use the name we still need to have parentheses to handle the comma
`(X := y,) == (y,)`

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: Generate a python parser with this branch and search for `,)` you'll see the examples of both of those cases

Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:333]: Now I get it. I just wished we could do it in such a way we do not generate extra parenthesis but I guess black can clean up for us. Thanks for bearing with me.

Comment (0dminnimda) [reply src/pegen/python_generator.py:333]: You're welcome!
Why do you not like generating parentheses in there?

Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:333]: The ones around the endmarker in the example above look funny to me but it is not a big deal.

**Comment by 0dminnimda** [comment]
@MatthieuDartiailh can you allow the workflow?

**Comment by MatthieuDartiailh** [comment]
Done

**Comment by MatthieuDartiailh** [comment]
@pablogsal would you have time to look at this ? or do you want to defer the final review to me maybe (since you seem quite busy ATM).

**Comment by pablogsal** [comment]
> @pablogsal would you have time to look at this ? or do you want to defer the final review to me maybe (since you seem quite busy ATM).

I apologize for the late reply @MatthieuDartiailh. Unfortunately the release of 3.11 is being ridiculously challenging and I'm expending all my OSS time on that :(

Please, go ahead and merge it if you feel confident. I trust your criteria :+1:

File: src/pegen/python_generator.py
Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py:399]: If you can address this comment @0dminnimda I will merge this PR. This the last blocker on my side and I got the green light from @pablogsal

Comment (0dminnimda) [reply src/pegen/python_generator.py:399]: It's not that big of a deal

Comment (MatthieuDartiailh) [reply src/pegen/python_generator.py]: I will add some comments to the different branches and merge when I get a chance.

**Comment by MatthieuDartiailh** [comment]
I wanted to add a couple small comments but apparently I cannot push to your repo. S I will merge and add the comment directly on main.

**Comment by 0dminnimda** [comment]
Weird usually there's no problem in that
Maybe i accidentally checked out Allow edits by maintainers..

**Pull request merged**
